compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt

```diff
@@ -0,0 +1,202 @@
+            /*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.codegen
+
+import com.intellij.util.PathUtil
+import com.intellij.util.SystemProperties
+import com.sun.jdi.ThreadReference
+import com.sun.jdi.VirtualMachine
+import com.sun.jdi.event.*
+import com.sun.jdi.request.EventRequest.SUSPEND_ALL
+import com.sun.jdi.request.StepRequest
+import com.sun.tools.jdi.SocketAttachingConnector
+import junit.framework.TestCase
+import org.jetbrains.kotlin.backend.common.output.OutputFile
+import org.jetbrains.kotlin.codegen.state.GenerationState
+import org.jetbrains.kotlin.test.ConfigurationKind
+import java.io.File
+import java.io.IOException
+import java.net.Socket
+import java.nio.file.Files
+import kotlin.properties.Delegates
+
+abstract class AbstractSteppingTest : CodegenTestCase() {
+    private companion object {
+        private const val DEBUG_ADDRESS = "127.0.0.1"
+        private const val DEBUG_PORT = 5225
+    }
+
+    override fun doMultiFileTest(wholeFile: File, files: List<TestFile>) {
+        createEnvironmentWithMockJdkAndIdeaAnnotations(ConfigurationKind.ALL, *listOfNotNull(writeJavaFiles(files)).toTypedArray())
+
+        val lineNumbers = wholeFile
+            .readLines()
+            .filter { it.matches("// \\d+.*".toRegex()) }
+            .single()
+            .drop(2)
+            .trim()
+            .split(" ")
+            .map { it.trim().toInt() }
+
+        loadMultiFiles(files)
+        val classBuilderFactory = OriginCollectingClassBuilderFactory(ClassBuilderMode.FULL)
+        val generationState = GenerationUtils.compileFiles(myFiles.psiFiles, myEnvironment, classBuilderFactory)
+        classFileFactory = generationState.factory
+
+        val tempDirForTest = Files.createTempDirectory("debuggerTest").toFile()
+
+        try {
+            val classesDir = File(tempDirForTest, "classes").apply {
+                for (classFile in classFileFactory.getClassFiles()) {
+                    File(this, classFile.relativePath).mkdirAndWriteBytes(classFile.asByteArray())
+                }
+            }
+
+            val process = startDebuggeeProcess(classesDir)
+            waitUntil { isPortOpen() }
```

Comment:
This is inherently unreliable and will block the test frequently. Can you find a random port to use? Also, this is probably not needed if you specify a timeout in `SocketAttachingConnector.attach`.

Reply:
Can't find a way to get a port and reserve it for later use, but definitely a valid point. Plan to address it in future.